File: Sources/BeagleUI/Sources/Widgets/UIComponents/ListView.swift
//
//  ListView.swift
//  BeagleUI
//
//  Created by Daniel Tes on 12/09/19.
//  Copyright © 2019 Daniel Tes. All rights reserved.
//

public struct ListView: Widget {
    
    let rows: [Widget]?
    let remoteDataSource: String?
    let loadingState: Widget?
    let direction: Direction
    
    init(
        rows: [Widget]? = nil,
        remoteDataSource: String? = nil,
        loadingState: Widget? = nil,
        direction: Direction = .vertical
    ) {
        self.rows = rows
        self.remoteDataSource = remoteDataSource
        self.loadingState = loadingState
        self.direction = direction
    }
}
extension ListView {
    public enum Direction {
        case vertical
        case horizontal
    }
}

File: framework/src/main/kotlin/br/com/zup/beagleui/framework/ui/ListView.kt
package br.com.zup.beagleui.framework.ui

import br.com.zup.beagleui.framework.core.Widget
import br.com.zup.beagleui.framework.layout.Horizontal
import br.com.zup.beagleui.framework.layout.Vertical
import br.com.zup.beagleui.framework.util.RowBuilder
import br.com.zup.beagleui.framework.util.generateRows

enum class ListDirection {
    VERTICAL,
    HORIZONTAL
}

data class ListView(
    val remoteDataSource: String? = null,
    val loadingState: Widget? = null,
    @Transient val size: Int,
    @Transient val rowBuilder: RowBuilder,
    @Transient val direction: ListDirection = ListDirection.VERTICAL
) : Widget {
    override fun buildResultName(): String = "rows"

    override fun build(): Widget {
        val children = generateRows(size, rowBuilder)

        return if (direction == ListDirection.VERTICAL) {
            Vertical(children = children)
        } else {
            Horizontal(children = children)
        }
    }
}